### ISSUE TITLE
Broadband FOMs

### ISSUE THREAD

**Issue by smartalecH** [facebookresearch/fmmax#23]
It would be great to be able to specify broadband FOMs (with optional weighting functions).

Since `fmmax` is a frequency domain method, one would either have to run each frequency solve sequentially, or distribute each solve in parallel (e.g. using MPI). It would be could to have a standard API for both options. The latter option, which may appear difficult, should actually be straightforward thanks to tools like [mpi4jax](https://github.com/mpi4jax/mpi4jax).

One thing to be aware of: If the specified FOM is scalar-valued (e.g. we take the mean across multiple frequencies) then jax will handle the gradients just fine (even if we use MPI). However, if we specify a *vector-valued* FOM (e.g. for performing a minimax optimization) then jax has some trouble propagating the gradients... there's some ways around this but it will require some tweaking.

**Comment by mfschubert** [comment]
This would indeed be useful. I am curious about the issue you are anticipating: would `jax.jacrev` not be adequate? Is it somehow incompatible with mpi4jax?

**Comment by smartalecH** [comment]
We had some issues with this when wrapping Meep's adjoint solver with jax. Jax had trouble compiling the graph when the FOM was vector-valued and we had a custom vjp specified... but I'm hopeful that this approach is different enough that we won't encounter the same issue.

**Comment by mfschubert** [comment]
OK, I did a quick check to see if we have an issue. I believe (fingers crossed) that we are safe, and `jacrev` will just work.

```python
from fmmax import utils

def my_fn(m):
    eigvals, eigvec = utils.eig(m)
    return jnp.concatenate([eigvals.flatten(), eigvec.flatten()])

m = jax.random.normal(jax.random.PRNGKey(0), (3, 3), dtype=complex)
jac = jax.jacrev(my_fn, holomorphic=True)(m)

print(jac.shape)  # (12, 3, 3)
```

Although I don't recall the details, I seem to recall having encountered issues similar to what you're describing. Maybe we are safe here because everything under the hood here is native jax.

**Comment by ianwilliamson** [comment]
> Jax had trouble compiling the graph when the FOM was vector-valued and we had a custom vjp specified... but I'm hopeful that this approach is different enough that we won't encounter the same issue.

Drive-by comment from me... if you look at the implementation of `jax.jacrev()`, it's just vmap-ing the pullback function created by `jax.vjp` over the basis vectors of the output. Indeed, a custom_vjp is not sufficient for enabling the use of `jax.vmap`. The contents of the custom VJP must consist entirely of JAX primitives with batching rules. Obviously, this was not the case with Meep, but it does seem to be the case with FMMAX!

Also... `jax.jacrev()` might not be particularly useful since it doesn't return the primal output (the vector-valued constraint in your case). You're probably better off doing the vmapping of the VJP yourself.

**Comment by mfschubert** [comment]
Thanks @ianwilliamson, yes this makes sense.

By the way, we could make use `jacrev` by simply returning the primal as an auxiliary quantity and using `has_aux=True`, as follows:

```python
def my_fn(m):
    eigvals, eigvec = utils.eig(m)
    y = jnp.concatenate([eigvals.flatten(), eigvec.flatten()])
    return (y, y)

m = jax.random.normal(jax.random.PRNGKey(0), (3, 3), dtype=complex)
jac, y = jax.jacrev(my_fn, holomorphic=True, has_aux=True)(m)

print(jac.shape)  # (12, 3, 3)
assert jnp.all(my_fn(m)[0] == y)
```

@smartalecH we can probably close this issue, and maybe create a new one calling for an example that exercises this functionality if that's appropriate.

**Comment by smartalecH** [comment]
Thanks @ianwilliamson for popping in! (I thought about tagging you earlier, but didn't know how busy you'd be... glad you stumbled you're way across this 😄 )

@mfschubert it might be good to keep this open until we either have an example or an API entry that can do broadband optimization both serially and with MPI. But I'm happy either way.
